Improvements and refactors to URL resolvers

NISH1001 opened the pull request:
# Refactor/improve URL resolvers

Now resolvers are their own subpackage at `akd.tools.resolvers`

### Summary :memo:
This PR refactors the URL resolver system by extracting resolver functionality from the scrapers module and organizing it into a dedicated `akd.tools.resolvers` subpackage. The refactoring improves code organization, separation of concerns, and provides a cleaner API for URL resolution operations.

### Details
_Describe more what you did on changes._
1. **Created dedicated resolvers subpackage** - Moved resolver classes from `akd/tools/scrapers/resolvers.py` to `akd/tools/resolvers/` with proper module organization
2. **Improved resolver architecture** - Enhanced base classes with better input/output schemas and configuration management
3. **Added specialized resolvers** - Implemented domain-specific resolvers for ADS, ArXiv, DOI, and PDF URL resolution
4. **Added Unpaywall resolver** - Integrated Unpaywall resolver from [Kumar's commit](https://github.com/NASA-IMPACT/accelerated-discovery/pull/120/commits/555271686ed1290a19f471eabb9e858ebec6d8da) for open access article resolution
5. **Created composite resolver** - Implemented `ResearchArticleResolver` that combines multiple resolvers for comprehensive URL resolution
6. **Updated factory integration** - Modified `akd/tools/factory.py` to use the new resolver subpackage imports
7. **Enhanced error handling and logging** - Improved resolver reliability with better error handling and debugging capabilities
8. **Maintained backward compatibility** - Ensured existing scraper functionality continues to work with the new resolver architecture

### Usage

The new resolver subpackage provides a clean API for URL resolution:

```python
from akd.tools.resolvers import (
    ResolverInputSchema,
    IdentityResolver,
    ArxivResolver,
    ResearchArticleResolver,
    PDFUrlResolver,
    DOIResolver,
    UnpaywallResolver,
)

# Basic usage with single resolvers
resolver = IdentityResolver()
result = await resolver.arun(ResolverInputSchema(url="https://example.com"))

# ArXiv-specific resolution
arxiv_resolver = ArxivResolver()
result = await arxiv_resolver.arun(
    ResolverInputSchema(url="https://arxiv.org/abs/2411.08181")
)

# Composite resolver for comprehensive URL resolution
composite_resolver = ResearchArticleResolver(
    PDFUrlResolver(),
    DOIResolver(),
    IdentityResolver(),
    debug=True,
)

# Handle multiple URL sources (URL, PDF URL, DOI)
result = await composite_resolver.arun(
    ResolverInputSchema(
        url="https://arxiv.org/abs/2411.08181",
        pdf_url="https://arxiv.org/abs/2411.08181",
        doi="10.48550/arXiv.2411.08181",
    )
)
```

For corssref-based DOI resolution (uses title and maybe author list to try to infer the doi)

```python
from akd.tools.resolvers import (
    ResolverInputSchema,
    CrossRefDoiResolver,
)

from akd.tools.resolvers.crossref_doi import CrossRefDoiResolverConfig

_cfg = CrossRefDoiResolverConfig(
    debug=True,
    author_priority_strategy="all",
    title_fuzzy_threshold=70,
    author_fuzzy_threshold=85,
    validate_resolved_url=True,
)

_resolver = CrossRefDoiResolver(_cfg)
_result = await _resolver.arun(
    ResolverInputSchema(
        url="https://arxiv.org/abs/2411.08181",
        title="INDUS: Effective and Efficient Language Models for science",
    )
)

_result.model_dump()
```

For Unpaywall-based open access resolution:

```python
from akd.tools.resolvers import (
    ResolverInputSchema,
    UnpaywallResolver,
)

from akd.tools.resolvers.unpaywall import UnpaywallResolverConfig

_cfg = UnpaywallResolverConfig(email="user@example.com", debug=True)
_resolver = UnpaywallResolver(_cfg)

_result = await _resolver.arun(
    ResolverInputSchema(
        url="https://doi.org/10.1371/journal.pone.0000308",
        title="Test paper",
    ),
)

_result.model_dump()
```

### Bugfixes :bug:
- Fixed resolver import paths in composite scraper to use new subpackage structure
- Improved URL resolution reliability with better error handling and fallback mechanisms

### Checks
- [x] Refactored resolvers into dedicated subpackage
- [x] Updated imports and dependencies
- [x] Maintained API compatibility
- [x] Enhanced resolver architecture with proper schemas
- [x] Added comprehensive resolver types (ADS, ArXiv, DOI, PDF, Identity, Composite)

NISH1001 committed:
Refactor/improve url resolvers

Now resolvers are their own subpackage at `akd.tools.resolvers`

NISH1001 committed:
Add tests for resolvers

NISH1001 committed:
Merge branch 'develop' of github.com:NASA-IMPACT/accelerated-discovery into feature/resolver-improvements

NISH1001 committed:
IMprove IO schema for resolvers

- SearchResultItem is now inheirted from Base IO schema which makes it
  useful for message passing to other components
- Now the Resolver Input and Ouputschema is direclty inheirted from
  SearchResultItem for better flexiblity

NISH1001 committed:
Add authors list to searchresult item direclty

NISH1001 committed:
Add DOI resolution for arxiv resolver

NISH1001 committed:
Refactor arxiv resolver

NISH1001 committed:
Retain parameters from input in resolvers that aren't mutated by teh
resolvers

This way we can can still return the original values for other params
that aren't changed

NISH1001 committed:
Merge branch 'develop' into feature/resolver-improvements

iamsims committed:
Add Component for Doi resolution using crossref

- Use title and author(if present) for querying crossref
- Fuzzy matches against the title
- Fuzzy matches against the top 3 title: author names use acronyms, and there are different number of authors for the same publication - which is why the default threshold is set low (70)

iamsims committed:
Add change in CrossRefDoiResolver for consistency with other resolver classes

- Remove arun method - now uses arun of the base class
- Change I/O type of resolve to SearchResultItem
- Force validate_url=False - to avoid _post_validate_url
- Extract authors from result item inside resolve and make authors optional in i/o type

iamsims committed:
Add accumluative composite resolver

- url and doi are populated using the resolvers
- resolver attribute in searchresultitem changed to list, and tests changed accordingly

iamsims committed:
Return doi by Crossrefdoiresolver if it already exist

iamsims committed:
Avoid calling resolvers if both url and doi are filled

iamsims committed:
Avoid replacing the original url by resolved url

- Modify composite accordingly
- Modify the schema in the test

NISH1001 commented on akd/tools/resolvers/crossref_doi.py:
```diff
@@ -0,0 +1,213 @@
+    def __init__(self, **kwargs):
+        kwargs["validate_resolved_url"] = False
+        super().__init__(**kwargs)
+
```
Don't need this because it's a pydantic schema. @iamsims

NISH1001 committed:
Use literal type hint for fuzzy match method

NISH1001 committed:
Add rapidfuzz to pyproject

NISH1001 commented on akd/tools/resolvers/crossref_doi.py:
```diff
@@ -0,0 +1,229 @@
+        title_norm = title.lower().strip()
+        input_authors_norm = [a.lower().strip() for a in (authors or [])][
+            :3
+        ]  # cap at 3
```
Probably not hard code. And get from config

NISH1001 commented on akd/tools/resolvers/crossref_doi.py:
```diff
@@ -0,0 +1,229 @@
+                    return None
+
+                # --- matching ---
+                for item in items[:10]:
```
No hard-coding please. Take from config

NISH1001 committed:
Improve crossrefdoi resolver with smarter author matching strategy

NISH1001 committed:
Remove constructor initialization for pydantic schema in
crossrefdoiresolverconfig

Instead use field_validator

NISH1001 committed:
Configure max candiates to process in crossrefdoi resolver

muthukumaranR commented on akd/tools/resolvers/crossref_doi.py:
```diff
@@ -0,0 +1,461 @@
+        if not authors:
+            return []
+
+        max_authors = self.config.max_authors_to_check
+        if max_authors == 0 or len(authors) <= max_authors:
+            return authors
+
+        strategy = self.config.author_priority_strategy
+
```
no need to use config since `AbstractBase.__set_attrs_from_config`

muthukumaranR commented on akd/tools/resolvers/crossref_doi.py:
```diff
@@ -0,0 +1,461 @@
+class CrossRefDoiResolverConfig(ArticleResolverConfig):
+    """Configuration for the CrossRef DOI resolver."""
+
+    validate_resolved_url: bool = Field(
+        default=False,
```
any reason this defaults to "False" when others are "True" ?

iamsims replied:
This is because crossref doesn't resolve url, so we don't want to validate, this and also i force this attribute to False

NISH1001 committed:
Merge branch 'develop' into feature/resolver-improvements

muthukumaranR committed:
add UnpaywallResolver for DOI-to-open-access resolution

- Implement DOI resolution via Unpaywall API
- Add email configuration for API compliance
- Include test coverage

(cherry picked from commit 555271686ed1290a19f471eabb9e858ebec6d8da)

NISH1001 committed:
Migrate unpaywall resolver to new structure

- Move UnpaywallResolver from akd/tools/scrapers/resolvers.py to akd/tools/resolvers/unpaywall.py
- Adapt to current BaseArticleResolver structure:
  * Update method signature: resolve(params: ResolverInputSchema) -> ResolverOutputSchema | None
  * Use ResolverOutputSchema with metadata preservation via params.model_dump()
  * Set resolved_url field instead of returning URL string
  * Return None on failure instead of original URL
  * Use async httpx instead of requests
  * Track resolver usage in resolvers list
- Add UnpaywallResolver to akd/tools/resolvers/__init__.py exports
- Update tests to new structure and import paths:
  * Test ResolverOutputSchema return type instead of string
  * Test resolved_url field and resolvers list
  * Verify None returns on resolution failure
  * Update imports from akd.tools.resolvers

Co-authored-by: muthukumaran R <mr0051@uah.edu>

NISH1001 committed:
Add unpaywallresolver config

`email` id is required

NISH1001 committed:
Update doi field for result in unpaywall

NISH1001 committed:
Merge branch 'feature/resolver-improvements' of github.com:NASA-IMPACT/accelerated-discovery into feature/resolver-improvements
